deltacat/compute/compactor/utils/io.py

```diff
@@ -285,3 +241,125 @@ def limit_input_deltas(
     logger.info(f"Input uniform delta count: {len(rebatched_da_list)}")
 
     return rebatched_da_list, hash_bucket_count, high_watermark, require_multiple_rounds
+
+
+def fit_input_deltas(
+    input_deltas: List[Delta],
+    cluster_resources: Dict[str, float],
+    hash_bucket_count: Optional[int],
+    deltacat_storage=unimplemented_deltacat_storage,
+) -> Tuple[List[DeltaAnnotated], int, HighWatermark, bool]:
+    """
+    This method tries to fit all the input deltas to run into the existing cluster. Contrary to
+    'limit_input_deltas', it will not fail if the current cluster cannot run the compaction reliably.
+    It is the responsibility of the caller to ensure that they pass enough resources for the job to execute.
+
+    Note: There is a possibility that individual file could be very large, which makes the deltas non uniform.
+    In such scenarios, it is advisable to allocate multiple vCPUs to the tasks to ensure parallelism.
+
+    Args:
+        input_deltas: The input deltas to be normalized.
+        cluster_resources: Total available resources in the cluster.
+        hash_bucket_count: The hash bucket count.
+        deltacat_storage: An implementation of the DeltaCAT storage interface.
+
+    Returns:
+        Tuple of list of annotated deltas, recommended hash bucket count, high watermark,
+            and whether multiple rounds are required (which is always False)
+    """
+    worker_cpus = int(cluster_resources["CPU"])
+    total_memory = float(cluster_resources["memory"])
+    high_watermark = HighWatermark()
+    annotated_input_da_list = []
+    delta_bytes = 0
+    total_files = 0
+
+    if not input_deltas:
+        raise AssertionError("No input deltas found!")
+
+    for delta in input_deltas:
+        manifest_entries = delta.manifest.entries
+        position = delta.stream_position
+
+        for entry in manifest_entries:
+            delta_bytes += entry.meta.content_length
+
+        total_files += len(manifest_entries)
+
+        high_watermark.set(
+            delta.locator.partition_locator,
+            max(position, high_watermark.get(delta.locator.partition_locator)),
+        )
+        delta_annotated = DeltaAnnotated.of(delta)
+        annotated_input_da_list.append(delta_annotated)
+
+    def estimate_size(content_length):
+        return (content_length * 1.0 / delta_bytes) * total_memory
+
+    # Assuming each CPU consumes equal amount of memory
+    min_delta_bytes = total_memory / worker_cpus
+    rebatched_da_list = DeltaAnnotated.rebatch(
+        annotated_deltas=annotated_input_da_list,
+        min_delta_bytes=min_delta_bytes,
+        estimation_function=estimate_size,
+    )
+
+    # Recommended hash buckets based on the POC results
```

Comment:
"POC results" seems a bit vague to an external user, maybe we should clarify what types of benchmarks were used to derive the memory to hash bucket count ratio?